### PR TITLE
♿ TetraLogical: P1 components accessibility review

### PR DESCRIPTION
**Pull request by TetraLogicalHelpdesk**
Suggested changes and additions based on the TetraLogical accessibility review commissioned by @nainar / @caroqliu

x-ref https://github.com/ampproject/amp.dev/pull/4972

### CODE REVIEW

**Comment by TetraLogicalHelpdesk** [comment]
Note that the find/replace change to add `lang="en"` to all relevant instances of the demos may have been overzealous in places. Apologies if it turns out to be excessive.

**Comment by nainar** [comment]
@caroqliu to kick of the review?

File: validator/validator.pb.go [hunk 1]
```diff
@@ -2527,7 +2527,7 @@ var xxx_messageInfo_HtmlFormat proto.InternalMessageInfo
 // NEXT AVAILABLE TAG: 40
 type TagSpec struct {
 	// The html_format field tells the validator for which html formats
-	// (ie: (<html ⚡> vs <html a4⚡>) this HTML TagSpec is allowed to validate.
+	// (ie: (<html ⚡lang="en"> vs <html a4⚡>) this HTML TagSpec is allowed to validate.
```
Comment (caroqliu) [inline validator/validator.pb.go:2530]: Please remove `lang="en"` from the comments in this doc.

File: validator/validator.proto [hunk 2]
```diff
@@ -538,7 +538,7 @@ message HtmlFormat {
 // NEXT AVAILABLE TAG: 40
 message TagSpec {
   // The html_format field tells the validator for which html formats
-  // (ie: (<html ⚡> vs <html a4⚡>) this HTML TagSpec is allowed to validate.
+  // (ie: (<html ⚡lang="en"> vs <html a4⚡>) this HTML TagSpec is allowed to validate.
```
Comment (caroqliu) [inline validator/validator.proto:541]: Same as above, please remove from comments.

**Review by caroqliu — CHANGES_REQUESTED** [review]
Thanks for your thorough changes @patrickhlauke! Left a few comments.

**Comment by patrickhlauke** [comment]
thanks @caroqliu ... yeah, the dangers of find/replace (also, on my Windows system at least, it's really tricky to tell visually when the ⚡ has a space after it or not ... should have caught that though). I'll do a second pass of this tonight to address the cases you spotted there

File: extensions/amp-lightbox-gallery/amp-lightbox-gallery.md [hunk 3]
```diff
@@ -75,9 +75,9 @@ were originally on. Currently, only the `type='slides'` carousel is supported.
   layout="responsive"
   type="slides"
 >
-  <amp-img src="image1" width="200" height="100"></amp-img>
-  <amp-img src="image1" width="200" height="100"></amp-img>
-  <amp-img src="image1" width="200" height="100"></amp-img>
+  <amp-img src="image1" width="200" height="100" alt="Image 1"></amp-img>
```
[the remark this answers is not in the document]
Comment (patrickhlauke) [reply extensions/amp-lightbox-gallery/amp-lightbox-gallery.md:78]: Only in this instance where it's a placeholder image, to be honest. Deemed it borderline acceptable, mainly due to the sheer number of these sorts of images without `alt` at all (and the fact that not all of the instances I spotted were easy to find/replace throughout). This sort of thing is certainly more of a stopgap solution - ideally, all examples should have appropriate/more real-world descriptive alternative text.

File: extensions/amp-auto-ads/amp-auto-ads.md [hunk 4]
```diff
@@ -37,7 +37,7 @@ constraints specified by the ad network. These constraints will limit:
   <li>The minimum distance that there should be between any adjacent ads</li>
 </ul>
 In addition to this, ads will only be inserted in locations on the page that do
-not cause an unacceptable re-flow (as determined by attemptChangeSize).
+not cause an unacceptable re-flow (as determined by `attemptChangeSize`).
```
[the remark this answers is not in the document]
Comment (patrickhlauke) [reply extensions/amp-auto-ads/amp-auto-ads.md:40]: yes, as i'm assuming that's code and should be backticked?

File: extensions/amp-sidebar/0.1/amp-sidebar.md [hunk 5]
```diff
@@ -216,10 +216,6 @@ Please see [this example file](https://github.com/ampproject/amphtml/blob/master
 
 `<amp-sidebar>` supports drilldown (nested) menus through a child component named [`<amp-nested-menu>`](../../amp-nested-menu/amp-nested-menu.md). With `<amp-nested-menu>`, `<amp-sidebar>` can support nesting one or more layers of submenus (and transition between them) as demonstrated by the following example:
 
-[tip type="note"]
```
[the remark this answers is not in the document]
Comment (patrickhlauke) [reply extensions/amp-sidebar/0.1/amp-sidebar.md:219]: it's unclear to me what actually happens (as i've not reviewed the logic behind `amp-nested-menu`). I assumed there was some intended menu-like behaviour being tied to `<li>` (allowing cursor keys to be used), which is what would not be appropriate. what is the actual outcome of doing what the tip suggests?

[the remark this answers is not in the document]
Comment (patrickhlauke) [reply extensions/amp-sidebar/0.1/amp-sidebar.md:219]: bringing up https://playground.amp.dev/?url=https%3A%2F%2Fpreview.amp.dev%2Fdocumentation%2Fcomponents%2Famp-sidebar.example.1.html%3Fformat%3Dwebsites&format=websites and manually removing the `<ul>` and `<li>` elements inside that nested menu seems to not have any tangible effect on the outcome (things are still turned into keyboard-focusable `role="button"` elements regardless.

if there was some intended extra cursor key behaviour, it's not there in the demo/broken?

File: extensions/amp-selector/amp-selector.md [hunk 6]
```diff
@@ -131,17 +134,9 @@ The attributes above behave the same way as they do on a standard HTML [`<select
...
-    `none` (default): The tab key changes focus between items in the `<amp-selector>`. The user must press enter or space to change the selection. Arrow keys are disabled.
-  </li>
-  <li>
-    `focus`: Tab key gives focus to `<amp-selector>`. The user navigates between items with the arrow keys. Must press space or enter to change the selection.
-  </li>
-  <li>
-    `select`: Tab key gives focus to `<amp-selector>`. The selection changes as the user navigates options with arrow keys.
-  </li>
-</ul>
+- `none` (default): The tab key changes focus between items in the `<amp-selector>`. The user must press enter or space to change the selection. Arrow keys are disabled.
```
[the remark this answers is not in the document]
Comment (patrickhlauke) [reply extensions/amp-selector/amp-selector.md:137]: note that currently, with the html there, it seems amp.dev doesn't then correctly render the backticks and shows them as literal content on the page - see https://amp.dev/documentation/components/amp-selector/#keyboard-select-mode

image

File: extensions/amp-sidebar/0.1/amp-sidebar.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (caroqliu) [reply extensions/amp-sidebar/0.1/amp-sidebar.md:219]: I'm not too sure, maybe @nainar can weigh in from having participated in design conversations? I assumed it was encouraging use of semantic HTML for a list of items as opposed to non-meaningful tags i.e. `<div>`, but if this does not improve UX, we don't need to encourage people to do this.

File: extensions/amp-auto-ads/amp-auto-ads.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (caroqliu) [reply extensions/amp-auto-ads/amp-auto-ads.md:40]: Sounds good, thanks for being so thorough :)

File: extensions/amp-3d-gltf/0.1/test/validator-amp-3d-gltf.out [hunk 7]
```diff
@@ -19,7 +19,7 @@ FAIL
 |    Tests for the amp-3d-gltf tag.
 |  -->
 |  <!doctype html>
-|  <html ⚡>
+|  <html ⚡lang="en">
```
[the remark this answers is not in the document]
Comment (patrickhlauke) [reply extensions/amp-3d-gltf/0.1/test/validator-amp-3d-gltf.out]: i'm currently unable to get a working local dev environment up and running for this (some quirk of windows/wsl/ubuntu that's beyond my technical ability to hunt down just now). is it ok to leave these as they are in my PR, and for you to re-update them properly at your end?

File: build-system/server/app-index/amphtml-helpers.js [hunk 8]
```diff
@@ -61,7 +61,7 @@ const AmpDoc = ({body, css, head, canonical}) => {
   assert(canonical);
   return html`
     <!DOCTYPE html>
-    <html ⚡>
+    <html ⚡lang="en">
```
[the remark this answers is not in the document]
Comment (patrickhlauke) [reply build-system/server/app-index/amphtml-helpers.js]: done

File: extensions/amp-ad-network-fake-impl/0.1/amp-ad-network-fake-impl.js [hunk 9]
```diff
@@ -127,7 +127,7 @@ export class AmpAdNetworkFakeImpl extends AmpA4A {
     const doc = new DOMParser().parseFromString(source, 'text/html');
     const root = doc.documentElement;
 
-    // <html ⚡> -> <html ⚡4ads>
+    // <html ⚡lang="en"> -> <html ⚡4ads>
```
[the remark this answers is not in the document]
Comment (patrickhlauke) [reply extensions/amp-ad-network-fake-impl/0.1/amp-ad-network-fake-impl.js]: done

File: validator/validator.pb.go (hunk 1 above)
Comment (patrickhlauke) [reply validator/validator.pb.go]: done

File: validator/validator.proto (hunk 2 above)
Comment (patrickhlauke) [reply validator/validator.proto]: done

**Review by banaag — APPROVED** [review]
LGTM as per validator related changes.

File: extensions/amp-lightbox-gallery/amp-lightbox-gallery.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (caroqliu) [reply extensions/amp-lightbox-gallery/amp-lightbox-gallery.md:78]: Makes sense, thank you

File: spec/amp-html-format.md [hunk 10]
```diff
@@ -170,7 +170,7 @@ In concrete terms this means that:
 AMP HTML documents MUST
 
 - <a name="dctp"></a>start with the doctype `<!doctype html>`. [🔗](#dctp)
-- <a name="ampd"></a>contain a top-level `<html ⚡>` tag (`<html amp>` is accepted as well). [🔗](#ampd)
+- <a name="ampd"></a>contain a top-level `<html ⚡ lang="en">` tag (`<html amp>` is accepted as well). [🔗](#ampd)
```
Comment (caroqliu) [inline spec/amp-html-format.md:173]: Please remove from this line, since `lang="en"` is not a "MUST"

File: src/auto-lightbox.js [hunk 11]
```diff
@@ -39,7 +39,7 @@ export const AutoLightboxEvents = {
  */
 export function installAutoLightboxExtension(ampdoc) {
   const {win} = ampdoc;
-  // Only enabled on single documents tagged as <html amp> or <html ⚡>.
+  // Only enabled on single documents tagged as <html amp> or <html ⚡ lang="en">.
```
Comment (caroqliu) [inline src/auto-lightbox.js:42]: Please remove `lang="en"` from this comment.

File: src/standalone.js [hunk 12]
```diff
@@ -23,7 +23,7 @@ import {isAmphtml} from './format';
  */
 export function installStandaloneExtension(ampdoc) {
   const {win} = ampdoc;
-  // Only enabled when the document is tagged as <html amp> or <html ⚡>.
+  // Only enabled when the document is tagged as <html amp> or <html ⚡ lang="en">.
```
Comment (caroqliu) [inline src/standalone.js:26]: Same as above

File: extensions/amp-a4a/0.1/test/testdata/valid_css_at_rules_amp.reserialized.js [hunk 13]
```diff
@@ -101,7 +101,7 @@ export const data = {
...
+  adTemplate: `<!doctype html><html ⚡ lang="en"><head><meta charset="utf-8"><link rel="canonical" href="self.html" /><meta name="viewport" content="width=device-width,minimum-scale=1"><style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript><script async src="https://cdn.ampproject.org/v0.js"></script><script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.1.js"></script>
```
Comment (caroqliu) [inline extensions/amp-a4a/0.1/test/testdata/valid_css_at_rules_amp.reserialized.js:104]: Let's revert this file -- the change appears to causes the following test files to fail:
- `extensions/amp-a4a/0.1/test/test-template-renderer.js` 
- `extensions/amp-a4a/0.1/test/test-template-validator.js` 
- `extensions/amp-ad-custom/0.1/test/test-amp-ad-custom.js`

@ampproject/wg-monetization Not too familiar with these files, is it expected for this kind of change to cause some test failures? Corresponding Travis run: https://travis-ci.com/github/ampproject/amphtml/jobs/443031330

**Review by caroqliu — CHANGES_REQUESTED** [review]

File: spec/amp-html-format.md (hunk 10 above)
Comment (TetraLogicalHelpdesk) [reply spec/amp-html-format.md]: done

File: src/standalone.js (hunk 12 above)
Comment (TetraLogicalHelpdesk) [reply src/standalone.js]: done

File: extensions/amp-a4a/0.1/test/testdata/valid_css_at_rules_amp.reserialized.js (hunk 13 above)
Comment (TetraLogicalHelpdesk) [reply extensions/amp-a4a/0.1/test/testdata/valid_css_at_rules_amp.reserialized.js]: done

File: src/auto-lightbox.js (hunk 11 above)
Comment (TetraLogicalHelpdesk) [reply src/auto-lightbox.js]: done

**Review by caroqliu — COMMENTED** [review]
Thanks so much for your diligence and thoroughness here! Looks like there are a couple merge conflicts - once those are resolved and tests are healthy I'd be happy to sign off.

**Comment by TetraLogicalHelpdesk** [comment]
> Thanks so much for your diligence and thoroughness here! Looks like there are a couple merge conflicts - once those are resolved and tests are healthy I'd be happy to sign off.

I've resolved the merge conflicts by resetting the validator/testdata files (as noted above, I'm currently unable to sort out a working build environment for some reason, with obscure WSL errors, so can't actually run the tests, nor generate the `.out` files correctly). assuming that the presence or not of `lang="en"` in the test files is not mission-critical for testing the actual functionality/running the test suite. If you feel these should be changed too, let me know and I can try and craft a more specific change, just affecting the `.html` files (and leave updating the `.out` files as an exercise that can hopefully be done at your end?)

File: extensions/amp-a4a/0.1/test/testdata/valid_css_at_rules_amp.reserialized.js [hunk 14]
```diff
@@ -101,7 +101,7 @@ export const data = {
...
+  adTemplate: `<!doctype html><html ⚡><head><meta charset="utf-8"><link rel="canonical" href="self.html" /><meta name="viewport" content="width=device-width,minimum-scale=1"><style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript><script async src="https://cdn.ampproject.org/v0.js"></script><script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.1.js"></script>
```
Comment (caroqliu) [inline extensions/amp-a4a/0.1/test/testdata/valid_css_at_rules_amp.reserialized.js]: Please also reset `minifiedTemplateCreative` in this same file.

**Review by caroqliu — CHANGES_REQUESTED** [review]
> I've resolved the merge conflicts by resetting the validator/testdata files (as noted above, I'm currently unable to sort out a working build environment for some reason, with obscure WSL errors, so can't actually run the tests, nor generate the .out files correctly). assuming that the presence or not of lang="en" in the test files is not mission-critical for testing the actual functionality/running the test suite. If you feel these should be changed too, let me know and I can try and craft a more specific change, just affecting the .html files (and leave updating the .out files as an exercise that can hopefully be done at your end?)

With regards to the build environment, what kind of errors are you seeing? Do you mind filing a follow up issue, or would you like help debugging more actively via Slack? cc @rsimha as FYI

As for the validator tests, the validator is stringent down to the character, so the addition of `lang=en` in an input file and not the output file (or vice versa) will cause tests to fail. For now, I'd recommend to not modify `.html` validator files if their corresponding`.out` files cannot be updated according to those change in the same PR. cc @banaag for another look in light of these removals/updates.

**Comment by amaltas** [comment]
Can you try updating tests (.out files) using the instructions at: https://github.com/ampproject/amphtml/tree/master/validator#usage

**Comment by TetraLogicalHelpdesk** [comment]
Sorry, only just revisiting this (as we're working on another phase of these reviews), and I see that this got stalled at the time and is now conflicting. Is anybody able to help get this updated and merged?

**Comment by caroqliu** [comment]
> Sorry, only just revisiting this (as we're working on another phase of these reviews), and I see that this got stalled at the time and is now conflicting. Is anybody able to help get this updated and merged?

@TetraLogicalHelpdesk I spent some time this morning taking a stab at resolving the merge conflicts -- that led to [this change](https://github.com/TetraLogical/amphtml/compare/tetralogical-documentation-phase1...caroqliu:fix-merge-conflicts?expand=1) which includes all the intermediary commits. I'm not confident that creating a secondary pull request into this working branch is the best way to go about it, but the most important part is the diff in the last two commits - one ([8c14d45](https://github.com/TetraLogical/amphtml/commit/8c14d4569ebb476443d0d8cde0778f56068e16cd)) which resolves merge conflicts after merging with fresh master, and the other ([545bfba](https://github.com/TetraLogical/amphtml/commit/545bfba9de1186a45ebdf16a38edcc72a2c1a2dd)) which persists one of the original changes to the file's new location.

Hopefully you can execute the merge with the following recommendations above as reference if needed. Otherwise please please find me (Caroline Liu) on the [AMP Slack](https://docs.google.com/forms/d/e/1FAIpQLSd83J2IZA6cdR6jPwABGsJE8YL4pkypAbKMGgUZZriU7Qu6Tg/viewform?fbzx=4406980310789882877) and we can hopefully resolve this in live time together.

Finally, one recommendation going forward could be to break this PR up so the majority of changes aren't stalled by recent changes in the codebase to a few others. The most obvious is to separate the `lang="en"` additions from the handful of documentation changes, but another could also be to fragment the set of `lang="en"` additions as well by a few directories in the repo at a time.

**Comment by patrickhlauke** [comment]
@patrickkettner not sure if this falls into your area, but wondering if you'd be able to help push this over the line? for context, this was done as part of an audit/review in november, and i've since been on other projects (so had no time allocated to nurse this PR further). i'm also far from being a wizard with git, as evidenced by my "let's just make it one giant PR" approach here.

**Comment by patrickhlauke** [comment]
After wrapping my head around this, I managed to work out what the actual steps required were. thank you @caroqliu for doing the hard work with to resolve the conflicts ... once I worked out that all that was needed was for me to merge that work back into our branch and then push it back up, it all went smoothly. Sorry for sounding the "not a git expert" alarm bell, @patrickkettner. I *think* this should now be mergeable.

**Comment by patrickkettner** [comment]
sorry I missed the ping, @patrickhlauke 🤦‍♂️ - great job!

**Comment by patrickhlauke** [comment]
> That's a whole lot of `lang` attributes

yes, sorry...the joy/pain of doing a fairly global find and replace :)

**Review by caroqliu — APPROVED** [review]

**Comment by TetraLogicalHelpdesk** [comment]
> Looks like this PR has fallen behind a few merge conflicts. Anything I can do to help get this through? @TetraLogicalHelpdesk @patrickhlauke

Hi @caroqliu, it's probably best at this stage (since the project relating to this was closed quite some time ago) if somebody on your side could shepherd this PR through to completion. Alternatively, we can add some extra time to the next AMP-related project to take care of this? Let us know how you'd like to proceed.

**Comment by Gregable** [comment]
@kristoferbaxter for triage, unclear what should be done with this old PR

**Pull request closed without merging**